tests/run-macros/i8007/Macro_3.scala

```diff
@@ -55,7 +55,7 @@ object Eq {
 
       case '{ $m: Mirror.SumOf[T] { type MirroredElemTypes = elementTypes }} =>
         val elemInstances = summonAll[elementTypes]
-        val eqSumBody: (Expr[T], Expr[T]) => Expr[Boolean] = (x, y) => {
+        def eqSumBody(x: Expr[T], y: Expr[T])(using Quotes): Expr[Boolean] = {
```

Comment:
> It seems to me that the need to thread through Quotes might be a source of errors in practice. It's not obvious why it's necessary.

This would have been avoided if we could have made `Expr` path dependent on the instance of `Quotes`. Because we do not have this we need to do runtime checks to catch the cases that did leak an expression outside of a splice. 

Even if we then un-extrude the quote, there is a performance penalty for doing it the wrong way as the owners might be wrong and would need to be changed. With this check we ensure that we have the correct ones at creation. 

If we ensure that we have the correct context we will also be able to rely on it to perform stronger yChecks in the reflection API. Instead of letting the error pass and be detected by the global yCheck.

> The need might come from staging. As macros are much more widely used than staging, I think it's worth optimizing more for macros.

It does not come from staging. Here it came due to a badly designed API. `eqSumBody` should have been a `def` from the start. Probably it shouldn't be a local method as well.

> Is it possible to achieve the same goal without complicating the user interface (I mean with only one Quotes)? That would be a big contribution.

The point of adding the `Quotes` is to use a single `Quotes` instance to crate `'x`/`'y` in `${eqSumBody('x, 'y)})` and use the same instance in `eqSumBody` to create `'{ $ordx == $ordy && ... }`. This would be clearer if the method was not defined locally. 

Or simply write it without the method as
```scala
val elements = Expr.ofList(elemInstances)
'{
  eqSum((x: T, y: T) => {
    $m.ordinal(x) == $m.ordinal(y) && 
    $elements.apply($m.ordinal(x)).asInstanceOf[Eq[Any]].eqv(x, y)
  })
}
```
The original code looks messy because it was not well designed.

Reply:
> It seems to me that the need to thread through Quotes might be a source of errors in practice. It's not obvious why it's necessary.

The need might come from staging. As macros are much more widely used than staging, I think it's worth optimizing more for macros.

Is it possible to achieve the same goal without complicating the user interface (I mean with only one Quotes)? That would be a big contribution.

Reply:
I also removed the old `scopeId` checks and they can be subsumed by these checks.